### ISSUE TITLE
Padding issue

### ISSUE THREAD

**Issue by krishnak** [kendryte/nncase#99]
I know from a previous issue, padding is done on CPU. I want to avoid that

My input size is 15x20 

if I add zero padding in my code it becomes

17x22

Now if I do a 3x3 convolution with stride=2 and padding=valid

I should get  ((17-3)/2) +1,((22-3/2)+1 = 8,10

I get this output but nncase is adding an extra padding operation which is unnecessarily adding overheads.  How to avoid it? You can refer to slim-320.kmodel it has this extra padding layer.

I tried to pad(1,2) and then crop(0,1) so that input becomes 17x23 that is also not working, it adds extra padding operation. Can you advise @sunnycase

**Comment by sunnycase** [comment]
You should pad to 18x22 beacause KPU only support stride=2 with floor so 17/2=7, that's why nncase add a pre pad.

**Comment by krishnak** [comment]
I think it doesn't work if padding is required in inner layers. See attached image, first image no padding applied in code.This padding operation is taking 5-6ms over head (my intention is to reduce this over head).  Second image padded as per your suggestion using tensorflow, it is resulting in additional operations as expected .

Before padding

![beforepadding](https://user-images.githubusercontent.com/1259447/74805741-111b4680-530a-11ea-8394-9866587cf0e3.png)

After padding

![afterpadding](https://user-images.githubusercontent.com/1259447/74805739-0f518300-530a-11ea-9759-2599a1bf9890.png)

**Comment by sunnycase** [comment]
Can you attach your tflite?

**Comment by krishnak** [comment]
[modelwithzeropadding.zip](https://github.com/kendryte/nncase/files/4223863/modelwithzeropadding.zip)
This is the file which pads 15x20 to 17x22 this results in further padding by K210

The next file is one which pads 15x20 ->17x22 -> 18x22 - the result is same as above, 1 extra padding operations in K210
[modelpaddedto18_22.zip](https://github.com/kendryte/nncase/files/4223930/modelpaddedto18_22.zip)

The final one is 15x20 -> 17x22 -> 19x22 -> cropped to > 18x22
[croppedto18_22.zip](https://github.com/kendryte/nncase/files/4223949/croppedto18_22.zip)

**Comment by sunnycase** [comment]
in modelpaddedto18_22.zip I did'nt see the 'extra' pad op, only pad from 15 to 16, that is right and can't be avoid.

**Comment by sunnycase** [comment]
And the origin model is also right. Because of KPU's limitation, you can't avoid the pad.

**Comment by krishnak** [comment]
https://github.com/kendryte/nncase/issues/89#issuecomment-583697080

Is it correct to say unless a feature map is having even number of rows and columns prior padding is redudant, KPU is going to pad it any way.

or else  why would a 15x20 which doesn't need padding on the dimension with size 15 for a convolution by 3x3 stride to as N-f/S +1 = 15-3/2 +1 =7 is getting padded by KPU to 16x20.

**Comment by sunnycase** [comment]
KPU always do pad(1,1)
15x20 for a convolution by 3x3 stride =2 gets padded to 17x22 by KPU, and strided slice to 7x10 (KPU only support output feature map size same as input or floor(input/2))

**Comment by krishnak** [comment]
Ok here is what I am doing
I zero pad a 30x40 input in code, stride=2,kernel =3x3 ,PADDING=valid

so input is 32x42

32-3/2+1 = 29/2+1 = floor(14.5)+1 = 15
42-3/2+1=39/2+1=floor(19.5)+1 = 20

For the above operation - there is NO explicit PAD operation on nncase graph. i.e no CPU involved

I zero pad this 15x20
now input is 17x22
convolution with 3x3 kernel, stride=2, padding=VALID
17-3/2+1 = 7+1 =8 
22-3/2+1 = floor(9.5)+1 = 10

BUT there is a PAD operation after the convolution by KPU in the 15x20 cae.

What I don't understand is why the first case is NOT having PAD and 15x20 is having PAD in KPU operation when both are similar operations

Please refer to the model  in modelwithzeropadding.zip which has this scenario. I am missing something in getting this understood.

**Comment by sunnycase** [comment]
I said floor(input/2)) not padded input / 2.
So 30x40 / 2 -> 15x20 there is no pad needed
15x20 / 2 -> 7x10 so if you want 8x10 a pad is needed.

**Comment by krishnak** [comment]
ok, I got it. if I want an output of 7x10 then there won't be a padding?

**Comment by sunnycase** [comment]
Yes

**Comment by krishnak** [comment]
thanks let me try to change the model

**Comment by krishnak** [comment]
input 15x20
no padding of input in code
Convolution 3x3,stride=2,padding=VALID
keras/tensorflowlite produces an output of 7x9 for this convolution and not 7x10

So I padded input to
15x22
Convolution 3x3,stride=2,padding=VALID
Keras/tensorflow produces output 7x10. KPU is still padding




[modelwith7_10.zip](https://github.com/kendryte/nncase/files/4230717/modelwith7_10.zip)

```
SUMMARY
INPUTS
0       Input_0 1x3x240x320
OUTPUTS
0       Identity        1x18x15x20
1       Identity_1      1x18x7x10
2       Identity_2      1x18x30x40
```
![image](https://user-images.githubusercontent.com/1259447/74934672-bfa7b000-540c-11ea-8680-539886bd03c2.png)

**Comment by krishnak** [comment]
Just to simplify debugging,

 I have created a 1 layer with input 15x20 -> padded to 15x22 in the code, however this is getting converted to 16x20 in KPU
[simple710.zip](https://github.com/kendryte/nncase/files/4230939/simple710.zip)

**Comment by krishnak** [comment]
15-3/2+1,20-3/2+1 as per tensor flow this gives to 7,9 without padding. But as per KPU 15x20 produces without padding 7,10 - this equation can never be satisfied

**Comment by sunnycase** [comment]
> 
> 
> Just to simplify debugging,
> 
> I have created a 1 layer with input 15x20 -> padded to 15x22 in the code, however this is getting converted to 16x20 in KPU
> [simple710.zip](https://github.com/kendryte/nncase/files/4230939/simple710.zip)

As I have said KPU always does pad(1,1), there is done in hardware and cannot be configured.
So if KPU do stride=2 it starts first conv window from the padded value but tensorflow with valid padding starts from valid values. So nncase must add another pad before the KPU pad and crop the padded conv result after KPU has done stride=2.
So both the prefix and surfix pads are necessary.

**Comment by krishnak** [comment]
I know this a catch22 situation, I will close this, may be in a future hardware versions this gets addressed.

**Comment by krishnak** [comment]
Just a work around - if the input image is 320x240 then this padding issue crops up. I have tried with input size 224x224 and this doesn't bring out any padding in kmodel
